From 69d88038eef15f9cf43c6b072b9bfd8c9131952b Mon Sep 17 00:00:00 2001
From: Han Wang <416810799@qq.com>
Date: Fri, 31 Aug 2018 10:27:28 +0200
Subject: [PATCH 66/68] CacheQuotaStrategy: Fix resource leak when reading
 cache quotas

08-29 22:36:17.992   971   978 W System  : A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
08-29 22:36:17.993   971   978 W System  : java.lang.Throwable: Explicit termination method 'close' not called
08-29 22:36:17.993   971   978 W System  : 	at dalvik.system.CloseGuard.open(CloseGuard.java:221)
08-29 22:36:17.993   971   978 W System  : 	at java.io.FileInputStream.<init>(FileInputStream.java:168)
08-29 22:36:17.993   971   978 W System  : 	at com.android.internal.os.AtomicFile.openRead(AtomicFile.java:147)
08-29 22:36:17.993   971   978 W System  : 	at com.android.server.storage.CacheQuotaStrategy.setupQuotasFromFile(CacheQuotaStrategy.java:301)
08-29 22:36:17.993   971   978 W System  : 	at com.android.server.usage.StorageStatsService$H.handleMessage(StorageStatsService.java:508)
08-29 22:36:17.993   971   978 W System  : 	at android.os.Handler.dispatchMessage(Handler.java:106)
08-29 22:36:17.993   971   978 W System  : 	at android.os.Looper.loop(Looper.java:193)
08-29 22:36:17.993   971   978 W System  : 	at android.os.HandlerThread.run(HandlerThread.java:65)
08-29 22:36:17.993   971   978 W System  : 	at com.android.server.ServiceThread.run(ServiceThread.java:44)

Change-Id: I56953ee49528190ad4210715c8f5ae3446896828
---
 .../java/com/android/server/storage/CacheQuotaStrategy.java   | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/services/core/java/com/android/server/storage/CacheQuotaStrategy.java b/services/core/java/com/android/server/storage/CacheQuotaStrategy.java
index 7a35bf7bba1..92cb8c3b544 100644
--- a/services/core/java/com/android/server/storage/CacheQuotaStrategy.java
+++ b/services/core/java/com/android/server/storage/CacheQuotaStrategy.java
@@ -53,6 +53,8 @@ import com.android.internal.util.FastXmlSerializer;
 import com.android.internal.util.Preconditions;
 import com.android.server.pm.Installer;
 
+import libcore.io.IoUtils;
+
 import org.xmlpull.v1.XmlPullParser;
 import org.xmlpull.v1.XmlPullParserException;
 import org.xmlpull.v1.XmlSerializer;
@@ -309,6 +311,8 @@ public class CacheQuotaStrategy implements RemoteCallback.OnResultListener {
             cachedValues = readFromXml(stream);
         } catch (XmlPullParserException e) {
             throw new IllegalStateException(e.getMessage());
+        } finally {
+            IoUtils.closeQuietly(stream);
         }
 
         if (cachedValues == null) {
-- 
2.17.1

